ecobee/beam_batch.py

```diff
@@ -0,0 +1,144 @@
+import argparse
+import logging
+import os
+from collections import namedtuple
+from datetime import datetime
+from time import time
+
+import apache_beam as beam
+from apache_beam.io import ReadFromText
+from apache_beam.io import WriteToText
+from apache_beam.options.pipeline_options import PipelineOptions
+from apache_beam.options.pipeline_options import SetupOptions
+
+DataPoint = namedtuple("DataPoint", ["time", "action"])
+Sequence = namedtuple("Sequence", ["seq_len", "datapoints"])
+
+timestamp_prefix = "TS:"
+timestamp_postfix = "GMT "
+action_prefix = "Action:"
+date_fmt = "%Y-%m-%d %H:%M:%S %z"
+
+
+def to_data_point(item):
+    """
+    item =  TS:2018-12-03 13:29:15 +0000 GMT, Action:F
+    and converts it to DataPoint (2018-12-03 13:29:15 +0000, F) with first value as time (prefix and postfix removed)
+    and second value as just the action (prefix removed).
+    """
+    entry = item[: len(item) - 1].split(",") if item.endswith("\n") else item.split(",")
+    t, action = (
+        entry[0][len(timestamp_prefix) : -len(timestamp_postfix)],
+        entry[1][len(action_prefix) + 1 :],
+    )
+    return DataPoint(t, action)
+
+
+def process(words, time_period, top_n_elems):
+    """
+    :param words: list of DataPoints, in increasing order of time.
+    :param time_period if two events are apart by this value of time, they are considered in separate sequence
+    :param top_n_elems: the number of top sequences to return
+    :return: list of sequences
+    """
+    all_sequences = []
+    prev_item = words[0]
+    cur_sequence_actions = [prev_item]
+    cur_seq_len = 0
+    for item in words[1:]:
+        delta_seconds = (
+            datetime.strptime(item.time, date_fmt)
+            - datetime.strptime(prev_item.time, date_fmt)
+        ).total_seconds()
+        if delta_seconds >= time_period:
+            if cur_seq_len > 0:
+                all_sequences.append(Sequence(cur_seq_len, cur_sequence_actions))
+            else:
+                all_sequences.append(Sequence(delta_seconds, [prev_item, item]))
+            cur_sequence_actions = [item]
+            cur_seq_len = 0
+        else:
+            cur_sequence_actions.append(item)
+            cur_seq_len += delta_seconds
+        prev_item = item
+
+    if len(all_sequences) == 0:
+        all_sequences.append(Sequence(cur_seq_len, cur_sequence_actions))
+    all_sequences_sorted = sorted(all_sequences, key=lambda x: x[0], reverse=True)
+    return [
+        (
+            all_sequences_sorted[i].seq_len,
+            [(d.time, d.action) for d in all_sequences_sorted[i].datapoints],
+        )
+        for i in range(0, min(top_n_elems, len(all_sequences_sorted)))
+    ]
+
+
+def remove_if_exists(file_names):
+    for file_name in file_names:
+        if os.path.exists(file_name):
+            os.remove(file_name)
+
+
+def run(argv=None, save_main_session=True):
```

Comment:
This function is doing more than one thing. Can you break it down in multiple functions? You can extract out Argument parser into its own function and just return the parsed arguments.

Reply:
Command/Query separation
Functions should either do something or answer something, but not both.

Reply:
done.